code/game/machinery/vendors/vending.dm

```diff
@@ -326,24 +368,72 @@
 	if(item_slot_check(user, I))
 		insert_item(user, I)
 		return
-	return ..()
+	. = ..()
+	if(tiltable && !tilted && I.force)
+		if(resistance_flags & INDESTRUCTIBLE)
+			// no goodies, but also no tilts
+			return
+		var/should_warn = world.time > last_hit_time + hit_warning_cooldown_length
+		last_hit_time = world.time
+		if(should_warn)
+			visible_message("<span class='warning'>[src] seems to sway a bit!</span>")
+			to_chat(user, "<span class='danger'>You might want to think twice about doing that again, [src] looks like it could come crashing down!</span>")
+			return
+
+		switch(rand(1, 100))
+			if(1 to 5)
+				freebie(user, 3)
+			if(6 to 15)
+				freebie(user, 2)
+			if(16 to 25)
+				freebie(user, 1)
+			if(26 to 75)
+				return
+			if(76 to 90)
+				tilt(user)
+			if(91 to 100)
+				tilt(user, crit = TRUE)
+
 
+/obj/machinery/economy/vending/proc/freebie(mob/user, num_freebies)
+	visible_message("<span class='notice'>[num_freebies] free goodie\s tumble[num_freebies > 1 ? "" : "s"] out of [src]!</span>")
 
+	for(var/i in 1 to num_freebies)
+		for(var/datum/data/vending_product/R in shuffle(product_records))
+
+			if(R.amount <= 0)
+				continue
+
+			var/dump_path = R.product_path
+			if(!dump_path)
+				continue
+			new dump_path(get_turf(src))
+			R.amount--
```

Comment:
Yeah, this logic is kind of similar to that used when a vendor is broken and gives some free stuff. I might not be able to clean it up entirely, but I can probably at least combine those two use-cases to share the same functionality.

Reply:
Yeah, I gave it a go but it felt like it'd end up even messier if I tried to combine the two procs (since one has to go through every single item, and this can give duplicates), so I think it's probably good as is.